src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/OnTypeFormattingHandler.cs

```diff
@@ -112,47 +128,53 @@ public async Task<TextEdit[]> HandleRequestAsync(DocumentOnTypeFormattingParams
+            // There's a chance that the virtual document is not up to date. This can occur if the virtual document
+            // was not synchronized at the time TryGetDocument was called. To circumvent this, we wait until we have
+            // proper synchronization before proceeding.
+            var virtualDocument = await GetSynchronizedVirtualDocumentAsync(
+                request.TextDocument.Uri, documentSnapshot, cancellationToken).ConfigureAwait(false);
+            if (virtualDocument is null)
             {
-                Character = request.Character,
-                Options = request.Options,
-                Position = projectionResult.Position,
-                TextDocument = new TextDocumentIdentifier() { Uri = projectionResult.Uri }
-            };
-
-            cancellationToken.ThrowIfCancellationRequested();
+                return null;
+            }
 
-            _logger.LogInformation($"Requesting formatting for {projectionResult.Uri}.");
+            var csharpSourceText = SourceText.From(virtualDocument.Snapshot.GetText());
+            var document = GenerateRoslynCSharpDocument(csharpSourceText, _vsHostServicesProvider);
+            var documentOptions = await GetDocumentOptionsAsync(request, document).ConfigureAwait(false);
 
-            var languageServerName = triggerCharacterKind.Value.ToContainedLanguageServerName();
-            var response = await _requestInvoker.ReinvokeRequestOnServerAsync<DocumentOnTypeFormattingParams, TextEdit[]>(
-                Methods.TextDocumentOnTypeFormattingName,
-                languageServerName,
-                formattingParams,
+            // We call into Roslyn's formatting service directly via external access, which circumvents needing
+            // to send a request to the C# LSP server.
+            var formattingChanges = await RazorCSharpFormattingInteractionService.GetFormattingChangesAsync(
+                document, typedChar: request.Character[0], projectionResult.PositionIndex, documentOptions,
                 cancellationToken).ConfigureAwait(false);
-            var textEdits = response.Result;
-
-            if (textEdits is null)
+            if (formattingChanges.IsEmpty)
             {
                 _logger.LogInformation("Received no results.");
                 return null;
             }
 
+            var textEdits = formattingChanges.Select(
+                change => Extensions.TextChangeExtensions.AsTextEdit(change, csharpSourceText)).ToArray();
             _logger.LogInformation($"Received {textEdits.Length} results, remapping.");
 
             cancellationToken.ThrowIfCancellationRequested();
 
-            var remappedTextEdits = await _documentMappingProvider.RemapFormattedTextEditsAsync(projectionResult.Uri, textEdits, request.Options, containsSnippet: false, cancellationToken).ConfigureAwait(false);
+            var remappedTextEdits = await _documentMappingProvider.RemapFormattedTextEditsAsync(
```

Comment:
So this is the method that ends up calling `ApplyFormattedEditsAsync`, and I think this is called over LSP too.. so maybe on type formatting was worse than I thought, and it was actually 3 LSP requests in a row, and this "only" takes us down to 2.

I logged https://github.com/dotnet/aspnetcore/issues/36319 to follow up at some point, but I _think_ this is an unnecessary hoop to jump through now.

Reply:
Nice catch! It does seem like we could potentially get rid of that call.